src/shmem_comm.h

```diff
@@ -101,6 +101,40 @@ shmem_internal_put_nb(shmem_ctx_t ctx, void *target, const void *source, size_t
     }
 }
 
+static inline
+void
+shmem_internal_put_signal(shmem_ctx_t ctx, void *target, const void *source, size_t len, 
+                          uint64_t *sig_addr, uint64_t signal, int pe, long *completion)
+{
+    int node_rank;
+
+    if (len == 0) {
+        shmem_internal_put_scalar(ctx, sig_addr, &signal, sizeof(uint64_t), pe);
+        return;
+    }
+
+    if (-1 != (node_rank = SHMEM_GET_RANK_SAME_NODE(pe))) {
+#if USE_MEMCPY
+        memcpy(target, source, len);
+        shmem_internal_put_scalar(ctx, sig_addr, &signal, sizeof(uint64_t), pe);
+#elif USE_XPMEM
+        shmem_transport_xpmem_put(target, source, len, pe, node_rank);
+        shmem_internal_put_scalar(ctx, sig_addr, &signal, sizeof(uint64_t), pe);
+#elif USE_CMA
+        if (len > shmem_internal_params.CMA_PUT_MAX) {
+            shmem_transport_put_signal((shmem_transport_ctx_t *)ctx, target, source, len, sig_addr, signal, pe, completion);
+        } else {
+            shmem_transport_cma_put(target, source, len, pe, node_rank);
+            shmem_internal_put_scalar(ctx, sig_addr, &signal, sizeof(uint64_t), pe);
```

Comment:
@jdinan Instead of `shmem_internal_put_scalar`, will `shmem_transport_xpmem/cma_put` make more sense? The `shmem_transport_xpmem/cma_put` internally uses memcpy/process_vm_writev.. so, using these would not require any fence, right?